Allow custom level names in log output and arg parsing.

allisonvacanti opened the pull request:
This patch adds the ability to set a custom name for each log level.

My only concern is that it adds a bit of overhead by introducing more mutex locks, but this is worthwhile for the improved readability in our application. C++17's `shared_mutex` would likely make this a non-issue, since the most frequent accessed are for reading only. But for now, this just locks the existing loguru mutex.

emilk commented:
Adding an extra mutex lock (as well as a `map` lookup) to each log line seems unnecessary. A simple array of `std::atomic<Text>` would probably be much faster.

May I ask what your use case is?

allisonvacanti commented:
This is for the vtk-m project (https://gitlab.kitware.com/vtk/vtk-m).

We log to a variety of different log levels (performance, memory allocation, memory transfer, execution details, etc) and integrate with other tools (VTK, ParaView) that are likely to share our logging system and need to define their own levels in the same log. So right now we have log levels defined as:

```
enum Level {
  // Loguru levels
  Off = loguru::Verbosity_OFF,
  Fatal = ...,
  Error = ...,
  Warn = ...,
  Info = ...,

  // Reserved for application use
  UserBegin = 1,
  UserEnd = 255,

  // Our log levels:
  Performance = 256,
  MemControl,
  MemExecution,
  MemTransfer,
  Communication,
  (etc),

  // Space for more verbose application levels
  UserBeginVerbose = 1024,
  UserEndVerbose = 2048
};
```

with the `User` and `UserVerbose` levels reserved for consumers of our library. Since we span a large, sparse set of log levels, the array approach would be wasteful for our needs.

An alternative would be to hide this functionality behind some `ifdef`s so users could opt in to it only if they find the performance hit worthwhile.

Just FYI, we already maintain forks of our third-party codes, including loguru, so if you don't want to merge this one in, it's no big deal. We can simply keep this patch active in our fork if it doesn't fit with the broader goals of loguru.

allisonvacanti commented:
Looks like I'd also have to refactor out the std::string usage in the header section.

emilk commented:
That is a lot of log levels!

One very slim way to integrate this in Loguru is to add a callback hook for getting the name for a log-level:

Loguru header:
``` C
typedef void (*verbosity_name_t)(Verbosity); // Return nullptr to use default name.

// Set to nullptr to use default names.
void set_verbosity_name_callback(verbosity_name_t callback);
``` 

Loguru implementation:
``` C
static verbosity_name_t s_verbosity_name_callback;
void set_verbosity_name_callback(verbosity_name_t callback)
{
	s_verbosity_name_callback = callback;
}

…
// Usage:

	auto custom_level_name = s_verbosity_name_callback ? s_verbosity_name_callback(verbosity) : nullptr;
	if (custom_level_name) {
		snprintf(level_buff, sizeof(level_buff) - 1, custom_level_name.c_str());
	} else if (verbosity <= Verbosity_FATAL) {
		…
```

If the callback is null the cost is almost zero, and the added code is very short so the impact for others users is minimal. This would leave it to you to handle how to look-up names in an efficient manner (may I suggest a hash table?).

allisonvacanti commented:
I think that's a fine compromise! I'll see about refactoring our code to support this approach and update this MR later this week.

allisonvacanti commented:
This should be more agreeable. I also added callbacks to translate names --> levels to allow custom names in the argument parser, and moved the existing name<-->level logic into the public functions `const char* get_verbosity_name(Verbosity)` and `Verbosity get_verbosity_from_name(const char*)`, which also check the callbacks.

emilk commented:
Sweet, looks good!
